system/Database/BaseBuilder.php

```diff
@@ -1719,6 +1726,252 @@ public function getWhere($where = null, ?int $limit = null, ?int $offset = 0, bo
         return $result;
     }
 
+    /**
+     * Compiles batch insert/upsert strings and runs the queries
+     *
+     * @throws DatabaseException
+     *
+     * @return false|int|string[] Number of rows inserted or FALSE on failure, SQL array when testMode
+     */
+    protected function batchExecute(string $renderMethod, ?array $set = null, ?bool $escape = null, int $batchSize = 100)
+    {
+        if ($set === null) {
+            if (empty($this->QBSet)) {
+                if ($this->db->DBDebug) {
+                    throw new DatabaseException('You must use the "set" method to update an entry.');
+                }
+
+                return false; // @codeCoverageIgnore
+            }
+        } elseif (empty($set)) {
+            if ($this->db->DBDebug) {
+                throw new DatabaseException('insertBatch()/upsertBatch called with no data');
+            }
+
+            return false; // @codeCoverageIgnore
+        }
+
+        $hasQBSet = $set === null;
+
+        $table = $this->db->protectIdentifiers($this->QBFrom[0], true, null, false);
+
+        $affectedRows = 0;
+        $savedSQL     = [];
+
+        if ($hasQBSet) {
+            $set = $this->QBSet;
+        }
+
+        for ($i = 0, $total = count($set); $i < $total; $i += $batchSize) {
+            if ($hasQBSet) {
+                $QBSet = array_slice($this->QBSet, $i, $batchSize);
+            } else {
+                $this->setBatch(array_slice($set, $i, $batchSize), '', $escape);
+                $QBSet = $this->QBSet;
+            }
+            $sql = $this->{$renderMethod}($table, $this->QBKeys, $QBSet);
+
+            if ($this->testMode) {
+                $savedSQL[] = $sql;
+            } else {
+                $this->db->query($sql, null, false);
+                $affectedRows += $this->db->affectedRows();
+            }
+
+            if (! $hasQBSet) {
+                $this->resetRun([
+                    'QBSet'  => [],
+                    'QBKeys' => [],
+                ]);
+            }
+        }
+
+        $this->resetWrite();
+
+        return $this->testMode ? $savedSQL : $affectedRows;
+    }
+
+    /**
+     * Allows key/value pairs to be set for batch inserts/upserts
+     *
+     * @param mixed $key
+     *
+     * @return $this|null
+     */
+    public function setBatch($key, string $value = '', ?bool $escape = null)
+    {
+        $key = $this->batchObjectToArray($key);
+
+        if (! is_array($key)) {
+            $key = [$key => $value];
+        }
+
+        $escape = is_bool($escape) ? $escape : $this->db->protectIdentifiers;
+
+        $keys = array_keys($this->objectToArray(current($key)));
+        sort($keys);
+
+        foreach ($key as $row) {
+            $row = $this->objectToArray($row);
+            if (array_diff($keys, array_keys($row)) !== [] || array_diff(array_keys($row), $keys) !== []) {
+                // batch function above returns an error on an empty array
```

Comment:
What is the batch function above?

Reply:
I copied all of this from setInsertBatch(). I changed: // batchExecute() function returns an error on an empty array